boot/bootutil/src/image_validate.c

```diff
@@ -78,6 +95,20 @@ bootutil_img_hash(struct image_header *hdr, const struct flash_area *fap,
         if (rc) {
             return rc;
         }
+#ifdef MCUBOOT_ENC_IMAGES
+        if (fap->fa_id == FLASH_AREA_IMAGE_1 && hdr->ih_flags & IMAGE_F_ENCRYPTED) {
+            /* FIXME: fails if header size is larger than blk_sz */
+            if (off < hdr->ih_hdr_size) {
+                idx = hdr->ih_hdr_size;
+                blk_off = 0;
+            } else {
+                idx = 0;
+                blk_off = (off - hdr->ih_hdr_size) & 0xf;
+            }
+            boot_encrypt(fap, (off + idx) - hdr->ih_hdr_size, blk_sz - idx,
+                    blk_off, &tmp_buf[idx]);
```

Comment:
This might be simpler if you generated the encrypting stream such that offset 0 from the beginning of image is 0, but don't start xor'ing that stream to turn the image into ciphertext until it reaches of image header.
The way you have it here is fine too.
I.e. I believe you could say here: boot_encrypt(fap, off, blk_sz - idx, blk_off, &tmp_buf[idx]);
blk_off computation would change slightly too. blk_off = off & 0xf, or something like that.